core/src/http_helpers.rs

```diff
@@ -47,33 +54,45 @@ pub async fn read_body(
 		return Err(GenericTransportError::TooLarge);
 	}
 
-	let first_chunk =
-		body.next().await.ok_or(GenericTransportError::Malformed)?.map_err(GenericTransportError::Inner)?;
+	futures_util::pin_mut!(body);
 
-	if first_chunk.len() > max_request_body_size as usize {
-		return Err(GenericTransportError::TooLarge);
-	}
+	let mut received_data = Vec::with_capacity(body_size as usize);
+	let mut is_single = None;
 
-	let first_non_whitespace = first_chunk.iter().find(|byte| !byte.is_ascii_whitespace());
+	while let Some(d) = body.data().await {
+		let data = d.map_err(|e| GenericTransportError::Inner(anyhow!(e.into())))?;
 
-	let single = match first_non_whitespace {
-		Some(b'{') => true,
-		Some(b'[') => false,
-		_ => return Err(GenericTransportError::Malformed),
-	};
+		if data.chunk().len() > max_request_body_size as usize {
```

Comment:
Nit: since we also validate below that `body_size > max_request`, would it be safe to remote this check from here? 🤔

Reply:
yeah, good catch this check is useless